File: security-assessment-dashboard/frontend/src/types/host-inventory.ts
export type HostType = "host" | "website" | "api" | "domain" | "ip";
export type HostState = "up" | "down" | "unknown";
export type NetworkProtocol = "tcp" | "udp";
export type PortState = "open" | "closed" | "filtered" | "open|filtered" | "closed|filtered" | "unfiltered";
export type TechnologyCategory = "web_server" | "database" | "language" | "framework" | "middleware" | "operating_system" | "other";
export type ObservationCategory = "network" | "web" | "tls" | "auth" | "configuration" | "os" | "other";

export const HOST_TYPE_OPTIONS: { value: HostType; label: string }[] = [
  { value: "host", label: "Host" },
  { value: "website", label: "Website" },
  { value: "api", label: "API" },
  { value: "domain", label: "Domain" },
  { value: "ip", label: "IP" },
];

export const HOST_STATE_OPTIONS: { value: HostState; label: string }[] = [
  { value: "up", label: "Up" },
  { value: "down", label: "Down" },
  { value: "unknown", label: "Unknown" },
];

export interface NetworkInterface {
  id: string;
  host_id: string;
  ip_address: string;
  version: string;
  mac_address: string | null;
  network: string | null;
  interface_name: string | null;
}

export interface Technology {
  id: string;
  host_id: string;
  service_id: string | null;
  name: string;
  vendor: string | null;
  version: string | null;
  category: TechnologyCategory;
  first_seen: string;
  last_seen: string;
}

export interface OperatingSystem {
  id: string;
  host_id: string;
  vendor: string | null;
  family: string | null;
  name: string;
  version: string | null;
  accuracy: number;
  source: string;
  first_seen: string;
  last_seen: string;
}

export interface Service {
  id: string;
  host_id: string;
  port: number;
  protocol: NetworkProtocol;
  state: PortState;
  service_name: string | null;
  product: string | null;
  vendor: string | null;
  version: string | null;
  extra_info: string | null;
  banner: string | null;
  first_seen: string;
  last_seen: string;
}

export interface ObservationEvidence {
  id: string;
  observation_id: string;
  source_tool: string;
  title: string | null;
  content: string | null;
  file_path: string | null;
  created_at: string;
}

export interface Observation {
  id: string;
  host_id: string | null;
  service_id: string | null;
  port: number | null;
  plugin: string | null;
  source: string;
  category: ObservationCategory;
  observation_type: string | null;
  title: string;
  detail: string | null;
  first_seen: string;
  last_seen: string;
  evidence: ObservationEvidence[];
}

export interface ExecutionHistoryEntry {
  execution_id: string;
  tool_name: string;
  target_value: string;
  is_new: boolean;
  created_at: string;
}

export interface HostSummary {
  id: string;
  target_id: string | null;
  assessment_id: string;
  hostname: string | null;
  fqdn: string | null;
  ipv4: string | null;
  ipv6: string | null;
  mac_address: string | null;
  host_type: HostType;
  state: HostState;
  first_seen: string;
  last_seen: string;
  service_count: number;
}

export interface HostDetail {
  id: string;
  target_id: string | null;
  assessment_id: string;
  hostname: string | null;
  fqdn: string | null;
  ipv4: string | null;
  ipv6: string | null;
  mac_address: string | null;
  mac_vendor: string | null;
  host_type: HostType;
  state: HostState;
  fingerprint: string;
  first_seen: string;
  last_seen: string;
  source_execution_id: string | null;
  network_interfaces: NetworkInterface[];
  services: Service[];
  technologies: Technology[];
  operating_systems: OperatingSystem[];
  observations: Observation[];
  execution_history: ExecutionHistoryEntry[];
}

export interface HostListParams {
  assessment_id?: string;
  target_id?: string;
  host_type?: HostType;
  state?: HostState;
  search?: string;
  sort_by?: string;
  sort_dir?: "asc" | "desc";
  page?: number;
  page_size?: number;
}

export interface SearchResult {
  kind: "host" | "service" | "technology" | "observation" | "finding";
  id: string;
  host_id: string | null;
  /** Lets a search result deep-link into the Assessment that discovered it (its "Assets
   * Discovered" tab / Findings tab) instead of a dedicated per-resource page. */
  assessment_id: string | null;
  label: string;
  detail: string | null;
}

export interface SearchResponse {
  query: string;
  hosts: SearchResult[];
  services: SearchResult[];
  technologies: SearchResult[];
  observations: SearchResult[];
  findings: SearchResult[];
}

File: security-assessment-dashboard/backend/schemas/host_inventory.py
"""Request/response schemas for the Host Inventory & Observation Engine.

Every resource here is read-only from the API's perspective — rows are only
ever written by :class:`backend.services.host_inventory_service.HostInventoryService`
as a side effect of a completed scan, never through a CRUD endpoint. Per
``.claude/CLAUDE.md``: purely descriptive, no severity/CVSS anywhere.
"""

import uuid
from datetime import datetime

from pydantic import BaseModel

from backend.models.enums import (
    HostState,
    HostType,
    NetworkProtocol,
    ObservationCategory,
    PortState,
    TargetType,
    TechnologyCategory,
)


class NetworkInterfaceRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    host_id: uuid.UUID
    ip_address: str
    version: TargetType
    mac_address: str | None
    network: str | None
    interface_name: str | None


class TechnologyRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    host_id: uuid.UUID
    service_id: uuid.UUID | None
    name: str
    vendor: str | None
    version: str | None
    category: TechnologyCategory
    first_seen: datetime
    last_seen: datetime


class OperatingSystemRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    host_id: uuid.UUID
    vendor: str | None
    family: str | None
    name: str
    version: str | None
    accuracy: int
    source: str
    first_seen: datetime
    last_seen: datetime


class ServiceRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    host_id: uuid.UUID
    port: int
    protocol: NetworkProtocol
    state: PortState
    service_name: str | None
    product: str | None
    vendor: str | None
    version: str | None
    extra_info: str | None
    banner: str | None
    first_seen: datetime
    last_seen: datetime


class ObservationEvidenceRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    observation_id: uuid.UUID
    source_tool: str
    title: str | None
    content: str | None
    file_path: str | None
    created_at: datetime


class ObservationRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    host_id: uuid.UUID | None
    service_id: uuid.UUID | None
    port: int | None
    plugin: str | None
    source: str
    category: ObservationCategory
    observation_type: str | None
    title: str
    detail: str | None
    first_seen: datetime
    last_seen: datetime
    evidence: list[ObservationEvidenceRead] = []


class ExecutionHistoryEntryRead(BaseModel):
    """One row of a host's (or observation's) execution history — who touched it, and when."""

    execution_id: uuid.UUID
    tool_name: str
    target_value: str
    is_new: bool
    created_at: datetime


class HostSummaryRead(BaseModel):
    """Lightweight shape for the Discovered Hosts list page — no nested collections."""

    model_config = {"from_attributes": True}

    id: uuid.UUID
    target_id: uuid.UUID | None
    assessment_id: uuid.UUID
    hostname: str | None
    fqdn: str | None
    ipv4: str | None
    ipv6: str | None
    mac_address: str | None
    host_type: HostType
    state: HostState
    first_seen: datetime
    last_seen: datetime
    service_count: int = 0


class HostDetailRead(BaseModel):
    """Full Discovered Host Details page shape: every nested collection."""

    model_config = {"from_attributes": True}

    id: uuid.UUID
    target_id: uuid.UUID | None
    assessment_id: uuid.UUID
    hostname: str | None
    fqdn: str | None
    ipv4: str | None
    ipv6: str | None
    mac_address: str | None
    mac_vendor: str | None
    host_type: HostType
    state: HostState
    fingerprint: str
    first_seen: datetime
    last_seen: datetime
    source_execution_id: uuid.UUID | None
    network_interfaces: list[NetworkInterfaceRead] = []
    services: list[ServiceRead] = []
    technologies: list[TechnologyRead] = []
    operating_systems: list[OperatingSystemRead] = []
    observations: list[ObservationRead] = []
    execution_history: list[ExecutionHistoryEntryRead] = []


class SearchResult(BaseModel):
    """One matched row in a global search, generic across resource kinds.

    ``assessment_id`` lets the frontend deep-link a result into the
    Assessment that discovered it (its "Assets Discovered" tab) rather
    than a dedicated per-resource page — inventory data is contextual to
    an assessment, not a standalone destination.
    """

    kind: str
    id: uuid.UUID
    host_id: uuid.UUID | None
    assessment_id: uuid.UUID | None
    label: str
    detail: str | None


class SearchResponse(BaseModel):
    query: str
    hosts: list[SearchResult] = []
    services: list[SearchResult] = []
    technologies: list[SearchResult] = []
    observations: list[SearchResult] = []
    findings: list[SearchResult] = []
